File: src/apply_force_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <gazebo_msgs/srv/apply_link_wrench.hpp>

class ApplyForceNode : public rclcpp::Node
{
public:
  ApplyForceNode()
  : Node("apply_force_node")
  {
    client_ = this->create_client<gazebo_msgs::srv::ApplyLinkWrench>("/apply_link_wrench");
    timer_ = this->create_wall_timer(
      std::chrono::seconds(10), std::bind(&ApplyForceNode::send_request, this));
  }

private:
  void send_request()
  {
    if (!client_->wait_for_service(std::chrono::seconds(1))) {
      RCLCPP_ERROR(this->get_logger(), "Service /apply_link_wrench is not available.");
      return;
    }

    auto request = std::make_shared<gazebo_msgs::srv::ApplyLinkWrench::Request>();
    request->link_name = "copterPIX::root_link";  // Ensure this link name matches your model
    request->reference_frame = "";  // Empty reference frame means inertial frame
    request->reference_point.x = 0.0;
    request->reference_point.y = 0.0;
    request->reference_point.z = -0.35;     // set the position of the m16 relative to root_joint
    request->wrench.force.x = -2684.0;      // change with the actual kickback force of the gun!
    request->wrench.force.y = 0.0;      
    request->wrench.force.z = 0.0;
    request->wrench.torque.x = 0.0;
    request->wrench.torque.y = 0.0;
    request->wrench.torque.z = 0.0;
    request->start_time.sec = 0;
    request->start_time.nanosec = 0;
    request->duration.sec = -1;  // Negative duration means apply indefinitely
    request->duration.nanosec = 0;

    auto future_result = client_->async_send_request(request);
    auto status = future_result.wait_for(std::chrono::seconds(2));

    if (status == std::future_status::ready) {
      RCLCPP_INFO(this->get_logger(), "Force applied successfully.");
    } else {
      RCLCPP_ERROR(this->get_logger(), "Failed to call service apply_link_wrench.");
    }
  }

  rclcpp::Client<gazebo_msgs::srv::ApplyLinkWrench>::SharedPtr client_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ApplyForceNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/apply_force_node.py
# sudo apt install ros-humble-gazebo-ros-pkgs        is absoltely necessary!!!


import rclpy
from rclpy.node import Node
from gazebo_msgs.srv import ApplyLinkWrench

class ApplyForceNode(Node):

    def __init__(self):
        super().__init__('apply_force_node')
        self.client = self.create_client(ApplyLinkWrench, '/apply_link_wrench')
        self.timer = self.create_timer(5, self.send_request)

    def send_request(self):
        if not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().error('Service /apply_link_wrench is not available.')
            return

        request = ApplyLinkWrench.Request()
        request.link_name = 'copterPIX::root_link'  # Ensure this link name matches your model
        request.reference_frame = ''  # Empty reference frame means inertial frame
        request.reference_point.x = 0.0
        request.reference_point.y = 0.0
        request.reference_point.z = -0.35  # Set the position relative to root_joint
        request.wrench.force.x = -2684.0  # Change with the actual kickback force of the gun
        request.wrench.force.y = 0.0
        request.wrench.force.z = 0.0
        request.wrench.torque.x = 0.0
        request.wrench.torque.y = 0.0
        request.wrench.torque.z = 0.0
        request.start_time.sec = 0
        request.start_time.nanosec = 0
        request.duration.sec = -1  # Negative duration means apply indefinitely
        request.duration.nanosec = 0

        future = self.client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None:
            if future.result().success:
                self.get_logger().info('Force applied successfully.')
            else:
                self.get_logger().error(f'Failed to apply force: {future.result().status_message}')
        else:
            self.get_logger().error('Failed to call service apply_link_wrench.')

def main(args=None):
    rclpy.init(args=args)
    node = ApplyForceNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
